File: boletim-ocorrencia/src/main/java/br/edu/utfpr/td/tsi/sistema/boletim/ocorrencia/dao/InMemoryReportDAO.java
package br.edu.utfpr.td.tsi.sistema.boletim.ocorrencia.dao;

import br.edu.utfpr.td.tsi.sistema.boletim.ocorrencia.csv.FileReaderStolenVehicles;
import br.edu.utfpr.td.tsi.sistema.boletim.ocorrencia.model.StolenVehicleReport;
import jakarta.annotation.PostConstruct;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;

import java.util.ArrayList;
import java.util.List;

@Component
public class InMemoryReportDAO implements ReportDAO {

    @Autowired
    private FileReaderStolenVehicles fileReaderStolenVehicles;
    List<StolenVehicleReport> incidentReports;


    @PostConstruct
    public void init(){
        incidentReports = new ArrayList<>(fileReaderStolenVehicles.getReports());
    }

    @Override
    public void register(StolenVehicleReport stolenVehicleReport) {
        incidentReports.add(stolenVehicleReport);
    }

    @Override
    public void update(StolenVehicleReport stolenVehicleReport) {
        StolenVehicleReport existingReport = this.readById(stolenVehicleReport.getId());

        existingReport.setOccurrenceDate(stolenVehicleReport.getOccurrenceDate());
        existingReport.setOccurrenceLocal(stolenVehicleReport.getOccurrenceLocal());
        existingReport.setOccurrenceDayTime(stolenVehicleReport.getOccurrenceDayTime());
        existingReport.setStolenVehicle(stolenVehicleReport.getStolenVehicle());
    }

    @Override
    public StolenVehicleReport readById(int id) {
        return incidentReports.stream()
                .filter(report -> report.getId().equals(id))
                .findFirst()
                .orElse(null);
    }

    @Override
    public List<StolenVehicleReport> readByCity(String city) {
        final List<StolenVehicleReport> listReports = new ArrayList<>();
        incidentReports.stream()
                .filter(report -> report.getOccurrenceLocal().getCity().equals(city))
                .forEach(listReports::add);
        return listReports;
    }

    @Override
    public List<StolenVehicleReport> readByDayTime(String dayTime) {
        final List<StolenVehicleReport> listReports = new ArrayList<>();

        incidentReports.stream()
                .filter(report -> report.getOccurrenceDayTime().equals(dayTime))
                .forEach(listReports::add);
        return listReports;
    }

    @Override
    public void remove(int id) {
        incidentReports.remove(incidentReports.indexOf(this.readById(id)));
    }

    @Override
    public List<StolenVehicleReport> readAll() {
        return incidentReports;
    }
}

File: boletim_python_v/occurance_report_system/occurance_report_system/occurance_report/dao/InMemoryReportDao.py
from typing import List
from .ReportDao import ReportDAO
from ..models import stolen_vehicle_report


class InMemoryReportDAO(ReportDAO):
    def __init__(self):
        self.incident_reports = []

    def register(self, stolen_vehicle_report: stolen_vehicle_report) -> None:
        self.incident_reports.append(stolen_vehicle_report)

    def read_by_id(self, id: int) -> stolen_vehicle_report:
        for report in self.incident_reports:
            if report.get_id() == id:
                return report
        return None

    def read_by_city(self, city: str) -> List[stolen_vehicle_report]:
        list_reports = []
        for report in self.incident_reports:
            if report.get_occurrence_local().get_city() == city:
                list_reports.append(report)
        return list_reports

    def read_by_day_time(self, day_time: str) -> List[stolen_vehicle_report]:
        list_reports = []
        for report in self.incident_reports:
            if report.get_occurrence_day_time() == day_time:
                list_reports.append(report)
        return list_reports

    def remove(self, id: int) -> None:
        report = self.read_by_id(id)
        if report:
            self.incident_reports.remove(report)

    def read_all(self) -> List[stolen_vehicle_report]:
        return self.incident_reports
